### ISSUE TITLE
[no-object-literal-type-assertion] can't assert as generic type

### ISSUE THREAD

**Issue by zheeeng** [palantir/tslint#4239]
Not sure it should be a bug, but indeed expect we can assert an `object` as generic type.

### Bug Report

- __TSLint version__:  3.1.3
- __TypeScript version__: 5.11.0
- __Running TSLint via__: VSCode

#### TypeScript code being linted

```ts
export function shallowCopy<O extends object> (o: O) {
  // [tslint] Type assertion on object literals is forbidden, use a type annotation instead.
  const to = {} as O

  for (const key in o) {
    if (!o.hasOwnProperty(key)) continue

    to[key] = o[key]
  }

  return to
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "no-object-literal-type-assertion": true
  }
}
```

**Comment by jurijzahn8019** [comment]
I get this error for a given function

```typescript
export function getObjectOrEmpty<T>(): T {
  return getSomeObject() || ({} as T);
}


const { name } = getObjectOrEmpty<{ name: string | null }>();
```

My problem is: how i can properly return an empty object? 
If i do not want to return null? 

In my case destructuring syntax will not work on nullable result

**Comment by JoshuaKGoldberg** [comment]
@jurijzahn8019 that seems like it is still buggy behavior. It turns into:

```typescript
return getSomeObject() || ({} as { name: string | null })
```

...which is a dangerous and incorrect type cast.

**Comment by jurijzahn8019** [comment]
@JoshuaKGoldberg Ok, i accept.

although I do not see any difference between:

```
const t = {}
const a = { name: undefined }
```
In my case I only want to use destructuring.
And I definitely ok if name is not set.

I will find other way to solve it.
Thanks :slightly_smiling_face:

**Comment by JoshuaKGoldberg** [comment]
Oh there is a difference between `{}`, `{ name: null }`, and `{ name: undefined }` in the type system with [`strictNullChecks`](https://basarat.gitbooks.io/typescript/docs/options/strictNullChecks.html) enabled. `null` isn't `undefined`.

Without strict null checks, there's still a difference between `{}` and the other options. Whether `name` exists on the object is something the type system cares about by default. If you don't care whether the object exists on the interface, you can use:

```typescript
interface IFoo {
    name?: string;
};
```

...where the `?` indicates that the property both can be `undefined` _and_ doesn't have to exist on the object.

Closing this issue as declined for now, but if there's a great deal of interest in this feature, we can always revisit!
